File: markdowner-filter/src/main/java/org/jesperancinha/parser/markdowner/filter/SBTFilter.java
package org.jesperancinha.parser.markdowner.filter;

import lombok.extern.slf4j.Slf4j;
import org.apache.logging.log4j.util.Strings;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;

/**
 * Filter to check if folder contains a Simple Build Tool project and keeps the project name in memory
 */
@Slf4j
public class SBTFilter extends ProjectFilter<Path> {

    private static final String NAME = "name";

    @Override
    public boolean test(Path path) {
        final boolean maybeSBTBuild = path.getFileName().toString().equals("build.sbt");
        if (maybeSBTBuild) {
            try (BufferedReader bufferedReader = new BufferedReader(new FileReader(path.toFile()))) {
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    if (hasName(line)) {
                        return true;
                    }
                }
            } catch (IOException e) {
                log.trace("Not SBT format!");
            }
        }
        return false;
    }

    @Override
    public String lastProjectName() {
        return this.lastProjectName;
    }

    private boolean hasName(String line) {
        final String[] split = line.split("\\s*:=\\s*");
        if (split.length == 2) {
            return isNameProperty(split);
        }
        return false;
    }

    private boolean isNameProperty(String[] split) {
        final String left = split[0].trim();
        if (left.equals(NAME)) {
            return isValueAString(split[1]);
        }
        return false;
    }

    private boolean isValueAString(String s) {
        final String right = s.trim();
        if (Strings.isNotEmpty(right)) {
            this.lastProjectName = right.substring(1, right.length() - 1);
            return true;
        }
        return false;
    }
}

File: markdowner-filter/src/main/kotlin/org/jesperancinha/parser/markdowner/filter/SBTFilter.kt
package org.jesperancinha.parser.markdowner.filter

import org.apache.logging.log4j.util.Strings
import org.slf4j.LoggerFactory
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.nio.file.Path

/**
 * Filter to check if folder contains a Simple Build Tool project and keeps the project name in memory
 */
class SBTFilter : ProjectFilter<Path>() {
    override fun test(path: Path): Boolean {
        val maybeSBTBuild = path.fileName.toString() == "build.sbt"
        if (maybeSBTBuild) {
            try {
                BufferedReader(FileReader(path.toFile())).use { bufferedReader ->
                    var line: String
                    while (bufferedReader.readLine().also { line = it } != null) {
                        if (hasName(line)) {
                            return true
                        }
                    }
                }
            } catch (e: IOException) {
                logger.trace("Not SBT format!")
            }
        }
        return false
    }

    override fun lastProjectName(): String? {
        return lastProjectName
    }

    private fun hasName(line: String): Boolean {
        val split: Array<String> = line.split("\\s*:=\\s*".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return if (split.size == 2) {
            isNameProperty(split)
        } else false
    }

    private fun isNameProperty(split: Array<String>): Boolean {
        val left = split.get(0).trim { it <= ' ' }
        return if (left == NAME) {
            isValueAString(split.get(1))
        } else false
    }

    private fun isValueAString(s: String): Boolean {
        val right = s.trim { it <= ' ' }
        if (Strings.isNotEmpty(right)) {
            lastProjectName = right.substring(1, right.length - 1)
            return true
        }
        return false
    }

    companion object {
        private val NAME: String = "name"
        val logger = LoggerFactory.getLogger(SBTFilter::class.java)
    }
}